### DISCUSSION TITLE
Go SDK: support modules own types being used as and with interfaces

### DISCUSSION THREAD

**Discussion by sipsma** [dagger/dagger#12366] [maintainers]
Follow-up to https://github.com/dagger/dagger/pull/6254

Right now there's a fairly annoying limitation in that in Go modules it's not possible to use an object you've defined in your module to implement an interface you've defined in your module.

This is due to the fact that interfaces are passed by ID by right now objects defined in a module are still passed by raw json in the context of themselves (some previous discussion with more context on why [here](https://github.com/dagger/dagger/pull/6167#discussion_r1411467418)).

So to support this I think we'll probably just want to bite the bullet and deal with the complication of passing an object by id within the module it's defined in.

---

EDIT: another variation of the same sort of problem here came up: https://discord.com/channels/707636530424053791/1199101642260365432/1199110908639707186

Essentially, we also want *interfaces defined in your own module* to be implementable by any other types (whether from deps or your own types).

This in particular may actually call for blown self-API codegen, otherwise the module can't get codegen for the interfaces it has defined.

**Comment by chrisjpalmer** [comment]
+1 on this issue :)

**Comment by chrisjpalmer** [comment]
I have the following use case... I have a builder module defined as follows:


```go

/// git@github.com/org/buildermodule/.dagger
type Task interface {
	DaggerObject
   Do(ctx context.Context) error
}

type Builder struct {
   ExtraTasks []Task
}

func (b* Builder) WithExtraTask(extraTask Task) *Builder {
	b.ExtraTasks = append(b.ExtraTasks, extraTask)
    return b
}

func (b *Builder) Build(ctx context.Context) error {
		// do normal build stuff

	 for _, t := range b.ExtraTasks {
			if err := t.Do(ctx); err != nil {
				return err
		   }
	}

	return nil
}
```

Then I have module in a project repo which consumes the builder module:

```go
/// git@github.com/org/myrepo/.dagger
type MyRepo struct {}

func (r *MyRepo) Build() error {
	return dag.BuilderModule().WithExtraTask((&ExtraTask{}).AsBuilderTask()).Build() // not possible
}

type ExtraTask struct {}

func (*ExtraTask) Do(ctx context.Context) error {
// .. extra task
}
```

Unfortunately the above does not work as currently the `.AsBuilderTask()` method is not generated for this type. Currently I get around this like so:

```go
/// git@github.com/org/myrepo/.dagger/main
type MyRepo struct {}

func (r *MyRepo) Build() error {
	return dag.BuilderModule().WithExtraTask(dag.ExtraTask().AsBuilderTask()).Build()
}

/// git@github.com/org/main/.dagger/extratask
type ExtraTask struct {}

func (*ExtraTask) Do(ctx context.Context) error {
// .. extra task
}
```

Would love to see this implemented as it would remove the need to have an extra module :)
